### PR TITLE
[styled-components] Upgrade to 4.0.0

### PR DESCRIPTION
**Pull request by LavaToaster**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/styled-components/styled-components/blob/develop/CHANGELOG.md#v400-beta0---2018-09-04
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by Igorbek — CHANGES_REQUESTED** [review]
styled-components@4 is not released yet, so I assume a good amount of people is still using 3.x and will be for a while.
So my first suggestion would be to keep supporting 3.x and have 4.x in a new version folder so that we can have them both at least for some time.

**Comment by Lavoaster** [comment]
I don’t see why v3 types need to run along the v4 types in this repo. It’ll
still be available on an older npm tag, and I don’t think there’s anything
that’ll require changes to the v3 typings after v4 is released.

If anything we can update the docs on styles components to direct people to
install the correct types/styled-components version.

File: types/styled-components/styled-components-tests.tsx [hunk 1]
```diff
@@ -4,14 +4,16 @@ import * as ReactDOMServer from 'react-dom/server';
...
+    createGlobalStyle,
     isStyledComponent,
     keyframes,
     ServerStyleSheet,
     StyleSheetManager,
     ThemeProps,
     ThemeProvider,
     withTheme,
+    ThemeConsumer,
+    ThemeConsumerComponent,
```
Comment (probablyup) [inline types/styled-components/styled-components-tests.tsx]: What's this? We don't have an export called "ThemeConsumerComponent"

**Review by quantizor — COMMENTED** [review]
Aside from @Igorbek's comment (which I agree with) things look good! It'd be smart to maintain the v3 types for a bit since I think there are still some active discussions around them and potentially typing improvements to be contributed.

File: types/styled-components/styled-components-tests.tsx (hunk 1 above)
Comment (Lavoaster) [reply types/styled-components/styled-components-tests.tsx]: It's a type export, not a component. I accidentally left it in the imports even though it's unused. I'll be removing it from there shortly.

**Comment by Lavoaster** [comment]
@Igorbek Convention in DT appears to have the latest version in the root and historical versions in a folder. As such, I've added a v3 folder.

**Comment by Lavoaster** [comment]
@Igorbek I'll happily look into @simonbuchan's work, however, I'm not comfortable reformatting the package in such a way that goes against convention in DT. If you take a look at react, jquery, and others you'll see this convention in practice.

**Comment by Lavoaster** [comment]
Our changes are pretty much the same, barring two things:

1. Deprecation notice on `withComponent`.
	- I'm not sure if we should be adding in a deprecation notice just yet as, going by the docs, it hasn't actually been deprecated yet, there's just a public signal that it is a candidate for deprecation pending the `as` prop working for everyone's use cases.
2. Moving some react return types over from `StyledComponentClass` to `StyledComponentType`.
	- I can do this, but to be honest I don't understand the change. Would you be able to explain it @simonbuchan.

**Comment by simonbuchan** [comment]
While the vX folder *is* DT convention (Check the readme), generally the root is the latest one, and the other versions are only kept if you want to keep updating then in parallel.

I'll check this out later, I've been experimenting quite a bit since that initial version and there are some changes that I think make sense for a major version update (e.g. replacing all the ThemedX types with defaults on the X type, simpler DefaultTheme and Interpolation handling, regular descriptive naming), but while I did figure out the inferrance issue there are still some sticking points I was trying to figure out (the big one is doing as safely)

**Comment by Lavoaster** [comment]
@simonbuchan I've given you push access to my DT fork, in case you want to make changes so we don't have to do this over multiple PR's.

Additionally, If you expand on what you're trying to do (maybe with examples) I might be able to spend some time figuring it out with you.

**Comment by probablyup** [comment]
@simonbuchan do you have time to look into this in the next couple days? I want to formally release v4 on Monday so if the additional work can't be done beforehand, I'd say let's just accept @Lavoaster's PR after the suggestions by @Igorbek are addressed and you can plan on a major refactor of the types after.

File: types/styled-components/index.d.ts [hunk 2]
```diff
@@ -17,10 +18,8 @@ export type ThemedStyledProps<P, T> = P & ThemeProps<T>;
 export type StyledProps<P> = ThemedStyledProps<P, any>;
 
 export type ThemedOuterStyledProps<P, T> = P & {
+    as?: React.ComponentType | keyof JSX.IntrinsicElements;
```
[the remark this answers is not in the document]
Comment (Lavoaster) [reply types/styled-components/index.d.ts]: You’re right. I was thinking about this earlier. I’ll look into it when I have a chance.

**Comment by Lavoaster** [comment]
@probablyup I won’t be addressing @Igorbek’s feedback due to it going against DT convention. But I do want to resolve @diegohaz’s comment first.

**Comment by probablyup** [comment]
> Deprecation notice on withComponent.

Leave it out for now 👍

File: types/styled-components/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lavoaster) [reply types/styled-components/index.d.ts]: Alright so after playing around with this for a bit I don't believe this is possible in typescript. So to be honest the best way around this would be to sign a variable and give it a type which contains the required props.

**Comment by Lavoaster** [comment]
Alright, I've loosed up a type which would have prevented `as` from working in general. 

Additionally Typescript afaik cannot infer types from props so we're unable to extend the props on a component with the `as` prop like we are able to do with `withComponent` function. I did look at other types but I wasn't able to see anything that would work for me. This is probably the "safely" that @simonbuchan was talking about. The only way to allow for this extension would be to give it an "any" type which would mean that TS would no longer highlight invalid props.

TL;DR Types are ready for 4.0 from my perspective, we can fix current issues like the above later.

**Comment by diegohaz** [comment]
@Lavoaster TS can infer types from components with this (I'm not sure if it's the best approach though):
```ts
export type ComponentToProps<T> = T extends
  | React.ComponentType<infer P>
  | React.Component<infer P>
  ? P
  : never;
```

I was able to do something similar here: https://github.com/diegohaz/reuse

**Comment by simonbuchan** [comment]
Changes seem fine so far to me, and I'm happy for this to go in instead of mine, but if you're curious to see the changes I was talking about, and want to make the potentially breaking changes on the major version bump, instead of later, I've pushed an update to [my branch](https://github.com/simonbuchan/DefinitelyTyped/blob/styled-components-v4/types/styled-components/v4/index.d.ts), but keep in mind it's not currently working, see further below.

There's a lot of changes in there, the relevant things to look at are:
- Docs, renaming, sorting the types, and generally trying to make the file easier to read for users trying to figure out why their code isn't working (a lot of this is me trying to figure out what is in here). You may disagree with some of the choices here, and I'm not certain all my descriptions are correct!
- Deprecating all the `ThemedFoo<Theme>` types that can be replaced with `Foo<Theme = DefaultTheme>` (i.e. generic parameter defaults)
- Some negative type tests (e.g. `$ExpectError`) to check for things like `any` inference.
- Some attempts at type trickery to make types stronger in some cases (recursive interpolation?), or the types easier to reason about (e.g. `RelaxEmptyTheme`), these ones need a lot more tests!

I would expect a lot more type testing to be added to nail down the weirder things like mixing `.attr()` with inference from style function arg types, etc...

I backed out the `StyledComponentType<>` change @Lavoaster was asking about, as that was what was causing the inference problem I was originally asking about. That was due to me using `React.ComponentType<>`, not `ComponentClass<>` due to the `forwardRef()` change, but for some reason TS refused to infer the props type if the return type included a generic function.

It's not ready to ship as is: as mentioned I've not yet tackled `as`, though @diegohaz's suggestion looks good, but further the changes I've made (presumably to the interpolation types), mean the style function props are not being inferred again, but this time in reverse, e.g. `styled.div``${(props: MyProps) => props.myProp}``` should have props `MyProps`, but instead it complains that MyProps is not assignable to the default `never` for interpolation. Ugh.

In any case, if you want to rip anything from that, style or code, or if you don't, feel free! I don't know for sure when I'll have time to spend on this, so I wouldn't wait for it.

**Comment by probablyup** [comment]
@andy-ms Can we get this merged? I approve it as a core maintainer of styled-components.

**Comment by Igorbek** [comment]
@probablyup I'm looking this right now. Give me an hour please.

**Review by Igorbek — APPROVED** [review]

**Comment by probablyup** [comment]
Thanks @Igorbek for the review and @Lavoaster for jumping on the implementation!
